### PR TITLE
Piper/alternate api for making copies of serializables

### PR DESCRIPTION
**Pull request by pipermerriam**
Alternate API approach to https://github.com/ethereum/pyrlp/pull/77

### What was wrong

TODO

### How was it fixed.

```python
changeset = header.build_copy(number=3)  # optionally provide `args/kwargs` initial changes.
changeset.state_root = b'\x01...'  # changset is mutable
if changeset.coinbase == b'\xff...'  # field access works.
    ...

new_header = changeset.commit()  # builds a new copy with all of the changes.

# changeset.number = 7  # this would throw an error as the changeset is decommissioned on commit.
```

### CODE REVIEW

**Comment by gsalgado** [comment]
Maybe I'm missing something, but why not something as simple as a `make_mutable() -> Serializable` and `make_immutable() -> None` methods, like:
```
new_header = header.make_mutable()
new_header.foo = 'bar'
mutating_fn(new_header)
new_header.make_immutable()
header = new_header
# trying to mutate now will fail
```

**Comment by pipermerriam** [comment]
> but why not something as simple as a `make_mutable() -> Serializable` and `make_immutable() -> None` methods

Because I don't want there to be *mutable* versions of the `Serializable` class.  Otherwise, we can end up with an instances that is mutable hanging around.  The `ChangSet` object cannot be used as an RLP object since `rlp.encode` won't accept it as a `sedes`.

This API (or maybe some alternate form of it) gives us strict boundaries between mutable and immutable objects.

**Comment by pipermerriam** [comment]
I think that I'm going to vote 👎 on using a context manager API based on.

- The use of the object from the context manager after the context has exited seems like an anti-pattern.
- Allowing field access on the changset object after the context has exited seems to make it more likely that people will mis-use the object.
- Using a `Changeset` object in place of an RLP object likely makes using type hints harder for us.

So, I'm thinking the following now.

```python
changeset = header.build_copy(*args, **kwargs)  # creates a `Changeset` object with optional overrides from the `args/kwargs`

# you can read fields
assert changeset.number == header.number
# you can write fields
changeset.number = 5

# you can build an RLP object
do_thing_that_needs_a_header(changeset.as_serializable())  # like `commit()` but it doesn't decommission the changeset.

# you can finalize your changes
new_header = changeset.commit()
```

**Comment by carver** [comment]
https://github.com/ethereum/py-evm/pull/581#discussion_r183314702

> * The use of the object from the context manager after the context has exited seems like an anti-pattern.

Sure, so here's another option: use the context manager and explicitly disable the changeset outside of the context (stopping all read/write access, like you might do with a file handle).

```py
with header.build_copy() as header_draft:
    header_draft.field = 1
    do_things_that_modify_header(header_draft)

    # you *must* commit in context, if you want to generate the rlp object
    new_header = header_draft.commit()
    
# these both crash, because you shouldn't use it out of context:
assert header_draft.field == 1
header_draft.field = 2
```

> * Allowing field access on the changset object after the context has exited seems to make it more likely that people will mis-use the object.

See above. Also, it's not *more* likely, since passing around a changeset without any context manager is just as easy to misuse.

> * Using a Changeset object in place of an RLP object likely makes using type hints harder for us.

The explicit type difference seems like a win to me!

Methods that want to mutate a changeset explicitly declare that they accept a changeset instead of an RLP object. Methods that never try to modify an object declare that they accept RLP objects.

**Comment by pipermerriam** [comment]
cc @carver like this? 
aa6250a

File: rlp/sedes/serializable.py [hunk 1]
```diff
@@ -86,40 +93,66 @@ def __init__(self, field):
...
+    def __enter__(self):
+        if self.__state__ == ChangesetState.INITIALIZED:
+            self.open()
+            return self
+        else:
+            raise ValueError("Cannot open ChangeSet which is not in the INITIALIZED state")
+
+    def __exit__(self, exc_type, exc_value, traceback):
+        if exc_type is None:
+            self.close()
```
Comment (carver) [inline rlp/sedes/serializable.py]: Maybe this should close even on an exception? I suppose a file would.

**Comment by carver** [comment]
Looks like we have to go with https://docs.python.org/3.5/library/enum.html#autonumber instead of `enum.auto()` which was added in 3.6.

File: rlp/sedes/serializable.py (hunk 1 above)
Comment (pipermerriam) [reply rlp/sedes/serializable.py]: Corrected

**Pull request merged**